Sources/PackageMetadata/PackageMetadata.swift

```diff
@@ -0,0 +1,147 @@
+
+        let search = { (error: Error?) in
+            self.indexAndCollections.findPackages(query) { result in
+                do {
+                    let packages = try result.get().items.map { $0.package }.map { Package(identity: $0.identity,
+                                                                                           location: $0.location,
+                                                                                           versions: $0.versions.map { $0.version },
+                                                                                           readmeURL: $0.readmeURL) }
+                    if packages.isEmpty, let error = error {
+                        // If the search result is empty and we had a previous error, emit it now.
+                        return callback(.failure(error))
+                    } else {
+                        return callback(.success(packages))
+                    }
+                } catch {
+                    return callback(.failure(error))
+                }
+            }
+        }
+
+        let fetchStandalonePackageByURL = {
+            guard let url = URL(string: query) else {
+                return search(nil)
+            }
+
+            do {
+                try withTemporaryDirectory(removeTreeOnDeinit: true) {
+                    let tempPath = $0.appending(component: url.lastPathComponent)
+                    do {
+                        let repositorySpecifier = RepositorySpecifier(url: url)
+                        try self.repositoryProvider.fetch(repository: repositorySpecifier, to: tempPath, progressHandler: nil)
+                        if self.repositoryProvider.isValidDirectory(tempPath), let repository = try self.repositoryProvider.open(repository: repositorySpecifier, at: tempPath) as? GitRepository {
+                            let branches = try repository.getBranches()
+                            let versions = try repository.getTags().compactMap { Version($0) }
+                            let package = Package(identity: .init(url: url),
+                                                  location: url.absoluteString,
+                                                  branches: branches,
+                                                  versions: versions,
+                                                  readmeURL: self.guessReadMeURL(baseURL: url, defaultBranch: try repository.getDefaultBranch()))
+                            return callback(.success([package]))
+                        }
+                    } catch {
+                        return search(error)
+                    }
+                }
+            } catch {
+                return search(error)
+            }
+        }
+
+        if isRegistryIdentity {
+            return self.registryClient.getPackageMetadata(package: identity, observabilityScope: observabilityScope, callbackQueue: DispatchQueue.sharedConcurrent) { result in
```

Comment:
Maybe `result.tryMap` will help simplify the code?

Reply:
> I guess this works for GitHub repos only?

Yep, and only those who use `README.md` as the file name. I think we should improve on this, but this matches what is being done by the currently shipping Xcode, so seems like a reasonable starting point.

Reply:
Hm, seems like a lateral move to me, tbh